modules/apps/user-groups-admin/user-groups-admin-web/src/main/resources/META-INF/resources/js/ViewUserGroupsManagementToolbarPropsTransformer.js

```diff
@@ -0,0 +1,26 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+export default function propsTransformer({portletNamespace, ...otherProps}) {
+	return {
+		...otherProps,
+		onActionButtonClick(event, {item}) {
+			if (item.data?.action === 'deleteUserGroups') {
+				{
+					eval(`${portletNamespace}deleteUserGroups()`);
+				}
```

Comment:
can't we just call `${portletNamespace}deleteUserGroups()`?
i'm surprised `eval` is allowed, maybe it's because we're using it in some ancient scripts (cc @wincent)

Reply:
Even better, move the function in here and see if other usages of the global function could work with a similar approach.

Reply:
@julien, I thought I tried that, but I would get a Babel parse error.
```
{ SyntaxError: /Volumes/Projects/liferay/master/portal/modules/apps/user-groups-admin/user-groups-admin-web/src/main/resources/META-INF/resources/js/ViewUserGroupsManagementToolbarPropsTransformer.js: Unexpected token, expected ";" (21:6)

  19 | 			if (item.data?.action === 'deleteUserGroups') {
  20 | 				{
> 21 | 					${portletNamespace}deleteUserGroups()
     | 					 ^
  22 | 				}
  23 | 			}
  24 | 		},
    at Object._raise (/Volumes/Projects/liferay/master/portal/modules/node_modules/@babel/parser/lib/index.js:763:17)
    at Object.raiseWithData (/Volumes/Projects/liferay/master/portal/modules/node_modules/@babel/parser/lib/index.js:756:17)
    at Object.raise (/Volumes/Projects/liferay/master/portal/modules/node_modules/@babel/parser/lib/index.js:750:17)
    at Object.unexpected (/Volumes/Projects/liferay/master/portal/modules/node_modules/@babel/parser/lib/index.js:8930:16)
    at Object.semicolon (/Volumes/Projects/liferay/master/portal/modules/node_modules/@babel/parser/lib/index.js:8912:40)
    at Object.parseExpressionStatement (/Volumes/Projects/liferay/master/portal/modules/node_modules/@babel/parser/lib/index.js:11888:10)
    at Object.parseStatementContent (/Volumes/Projects/liferay/master/portal/modules/node_modules/@babel/parser/lib/index.js:11484:19)
    at Object.parseStatement (/Volumes/Projects/liferay/master/portal/modules/node_modules/@babel/parser/lib/index.js:11348:17)
    at Object.parseBlockOrModuleBlockBody (/Volumes/Projects/liferay/master/portal/modules/node_modules/@babel/parser/lib/index.js:11930:25)
    at Object.parseBlockBody (/Volumes/Projects/liferay/master/portal/modules/node_modules/@babel/parser/lib/index.js:11916:10)
  loc: Position { line: 21, column: 6 },
  pos: 800,
  code: 'BABEL_PARSE_ERROR' }
Command /Volumes/Projects/liferay/master/portal/modules/node_modules/@liferay/npm-scripts/node_modules/.bin/babel --no-babelrc --config-file /Volumes/Projects/liferay/master/portal/modules/apps/user-groups-admin/user-groups-admin-web/build/npmscripts/tmp/.babelrc src/main/resources/META-INF/resources --out-dir build/node/packageRunBuild/resources --source-maps exited with code 1
Received 1
error Command failed with exit code 1.
```

Reply:
Makes sense, I didn't see that when looking at that line.

Reply:
@jbalsas, moving the function there would also encounter the same issue as the function calls another namespaced function.  It would be:
```
const form = document.getElementById(
	`${portletNamespace}fm`
);

if (!form) {
	return;
}

${portletNamespace}doDeleteUserGroup(
	userGroupName,
	Liferay.Util.listCheckedExcept(
		form,
		`${portletNamespace}allRowIds`
	)
);
```

Reply:
You would need to do `window['${portlet.namesapce}deleteUsersGroups']()`

Reply:
Move the functions to `js` files and export them from there as we've done in other migrations. To use from the `propsTransformer` should be as simple as `import`ing them.

For the other case, we should check if we could migrate to a tag that also allows `propsTransformer` like `clay:dropdown` since that would allow us to reuse the same definition. If not, we could write the minimal code to maybe make it global (because I think its usage is set in Java) but without having the body of the function declared in the `jsp` file.

Reply:
> Attached are the changes I made for the user-groups-admin/user-groups-admin-web. I did not change the JS block in https://github.com/jonmak08/liferay-portal/blob/c40cc4e36c7bb4bf3c3c4cdd6d881739ad04dd2c/modules/apps/user-groups-admin/user-groups-admin-web/src/main/resources/META-INF/resources/view.jsp#L79...L191 like we mentioned. I just added a call to those methods in the PropsTransformer.js file for the MT which can be found at c40cc4e#diff-29e6a6e0815ac0587bad445103f881aeba2efd9b1a8705a7020fb76b2f89fd21.

I see you mentioned that, sorry I missed it, jumped straight to Julien's comment.

Let's try to move it and see what good things we have in place for it and what are we still missing.

Reply:
> i'm surprised `eval` is allowed, maybe it's because we're using it in some ancient scripts (cc @wincent)

We don't explicitly _disallow_ it, we just [use the `eslint:recommended` settings ](https://github.com/liferay/liferay-frontend-projects/blob/3140da045347c4577c7677a664eb5711ba85544b/projects/eslint-config/index.js#L14), and [those](https://eslint.org/docs/rules/) don't include [the `no-eval` rule](https://eslint.org/docs/rules/no-eval). I'd be surprised if we have many instances of `eval` out there though, and I'd be in favor of turning the rule on.

Reply:
:+1: Thanks for the info @wincent

Reply:
After some discussion with @jbalsas, because of all the inline JSP calls, we decided we will take only a minimalist approach to this migration and test using the initial suggestion of:

> You would need to do window['${portlet.namesapce}deleteUsersGroups']()

We will send that off the product team notifying them that is the best we can do with their code at the moment and they/us will have to refactor the code at a later time.

Reply:
Cool, and I created https://github.com/liferay/liferay-frontend-projects/issues/432 about turning the `no-eval` rule on by default.

Reply:
Thanks @wincent